hashes/src/macros.rs

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Public macros.
+
+/// Adds trait impls to a bytelike type.
+///
+/// Implements:
+///
+/// * `str::FromStr`
+/// * `fmt::{LowerHex, UpperHex}` using `hex-conservative`.
+/// * `fmt::{Display, Debug}` by calling `LowerHex`
+/// * `serde::{Deserialize, Serialize}`
+/// * `AsRef[u8; $len]`
+/// * `AsRef[u8]`
+/// * `Borrow<[u8; $len]>`
+/// * `Borrow<[u8]>`
+/// * `slice::SliceIndex<[u8]>`
+///
+/// Arguments:
+///
+/// * `ty` - The bytelike type to implement the traits on.
+/// * `$len` - The number of bytes this type has.
+/// * `$reverse` - `true` if the type should be displayed backwards, `false` otherwise.
+/// * `$gen: $gent` - generic type(s) and trait bound(s).
+#[macro_export]
+macro_rules! impl_bytelike_traits {
+    ($ty:ident, $len:expr, $reverse:expr $(, $gen:ident: $gent:ident)*) => {
+        impl<$($gen: $gent),*> $crate::_export::_core::str::FromStr for $ty<$($gen),*> {
+            type Err = $crate::hex::HexToArrayError;
+
+            fn from_str(s: &str) -> $crate::_export::_core::result::Result<Self, Self::Err> {
+                use $crate::hex::FromHex;
+
+                let mut bytes = <[u8; $len]>::from_hex(s)?;
+                if $reverse {
+                    bytes.reverse();
+                }
+                Ok(Self::from_byte_array(bytes))
+            }
+        }
+
+        $crate::hex::impl_fmt_traits! {
+            #[display_backward($reverse)]
+            impl<$($gen: $gent),*> fmt_traits for $ty<$($gen),*> {
+                const LENGTH: usize = $len;
+            }
+        }
+
+        $crate::serde_impl!($ty, $len $(, $gen: $gent)*);
+
+        impl<$($gen: $gent),*> $crate::_export::_core::convert::AsRef<[u8; $len]> for $ty<$($gen),*> {
+            #[inline]
+            fn as_ref(&self) -> &[u8; $len] { self.as_byte_array() }
+        }
+
+        impl<$($gen: $gent),*> $crate::_export::_core::convert::AsRef<[u8]> for $ty<$($gen),*> {
+            #[inline]
+            fn as_ref(&self) -> &[u8] { &self[..] }
+        }
+
+        impl<$($gen: $gent),*> $crate::_export::_core::borrow::Borrow<[u8; $len]> for $ty<$($gen),*>  {
+            fn borrow(&self) -> &[u8; $len] {  self.as_byte_array() }
+        }
+
+        impl<$($gen: $gent),*> $crate::_export::_core::borrow::Borrow<[u8]> for $ty<$($gen),*>  {
+            fn borrow(&self) -> &[u8] {  &self[..] }
+        }
```

Comment:
Maybe we should have `Mut` versions too. At least with some additional macro option.

Reply:
Agreed, but I do think it should be optional (and for hashes, off). I'd be fine adding in a followup. I guess, whatever followup moves the macro and makes it more generally useable.

Reply:
> and for hashes, off

Why? I think it should be only off for things like `Sighash`. I agree it can be in followup. The reason I haven't acked is the `SliceIndex` stuff.

Reply:
I guess it's harmless, but why would anybody want to mutate a hash?

Good observation though that with `Sighash` there *would* be meaningful harm, so there's a reason to go out of our way to avoid it.

Reply:
> why would anybody want to mutate a hash?

I really don't know except for `mem::swap`, but that'd just swap the entire hash without our support. (Though swapping with something weakly typed doesn't sound too crazy.)

My reason is mainly basic Rust consistency that unless we have a good reason to avoid a trait impl we should have the impl. This is also one of the Rust API guidelines.

Reply:
Cool. Sounds good to me.